vnext/Microsoft.ReactNative/Utils/BatchingEventEmitter.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "BatchingEventEmitter.h"
+#include "DynamicWriter.h"
+#include "JSValueWriter.h"
+
+namespace winrt::Microsoft::ReactNative {
+
+BatchingEventEmitter::BatchingEventEmitter(Mso::CntPtr<const Mso::React::IReactContext> &&context) noexcept
+    : m_context(std::move(context)) {
+  m_uiDispatcher = m_context->Properties().Get(ReactDispatcherHelper::UIDispatcherProperty()).as<IReactDispatcher>();
+}
+
+void BatchingEventEmitter::EmitJSEvent(int64_t tag, winrt::hstring &&eventName, JSValueObject &&eventObject) noexcept {
+  VerifyElseCrash(m_uiDispatcher.HasThreadAccess());
+
+  implementation::BatchedEvent newEvent{tag, std::move(eventName), std::move(eventObject)};
+
+  {
+    std::unique_lock lock(m_eventQueueMutex);
+
+    m_eventQueue.push_back(std::move(newEvent));
+
+    // Once a frame is presented, send a task to JS to emit all elements in the
+    // queue. We know the task is pending on the UI thread or JS thread if we
+    // have a non-zero size queue, so we want to only trigger pumping when the
+    // first element is added to an empty queue.
+    if (m_eventQueue.size() == 1) {
+      m_renderingRevoker = xaml::Media::CompositionTarget::Rendering(
+          winrt::auto_revoke, [weakThis{get_weak()}](auto const &, auto const &) {
+            if (auto strongThis = weakThis.get()) {
+              strongThis->OnFrameUI();
+            }
+          });
+    }
+  }
+}
+
+void BatchingEventEmitter::EmitCoalescingJSEvent(
+    int64_t tag,
+    winrt::hstring &&eventName,
```

Comment:
I know you've removed it for now, but a coalescing key will be useful once we have pointer move events potentially associated with different pointers.

Reply:
Yes, I agree. Going to leave this as followup for when we do touchMove.